docs/src/modules/java/pages/ai-coding-assistant-guidelines.adoc

```diff
@@ -4,43 +4,111 @@ include::ROOT:partial$include.adoc[]
+* Prefer a domain model prefix on classes defined within the `application` or `domain` packages to differentiate. For example, an application that contains both a shopping cart and an order would be expected to have the `ShoppingCartEvent` and `OrderEvent` records.
+* Never use unqualified class names like `Event`, `Workflow`, or `Agent`, unless nested inside another descriptive class or interface.
+* A domain record should not emit effects, which belong to the `application` package.
+* A domain record should be immutable.
+** Make a copy of collections when updating them.
+** To make single-field mutation easier, prefer the functional builder-style syntax, e.g. `withLocation(x)` that creates a new instance with the new field value rather than explicitly setting the location field
+
 
 == Entities
 
-- Commands can be defined as records inside the entity and they don't have to have a common interface.
-- Command handlers should be implemented in the entity and not in the domain object. The domain object should have methods for business logic validations but it should not handle commands directly or be concerned with entity effects.
-- Command handlers that make updates without returning any information should return `akka.Done` for successful responses and `effects().error()` for validation errors. Use static import for `akka.Done.done()` method.
-- Events should be defined in the domain package and the events for an entity should have a common sealed interface, and define `@TypeName` for serialization.
-- `applyEvent` method should never return null, return the current state or throw an exception.
-- Key Value Entities, Event Sourced Entities, Workflows can accept only single method parameter, wrap multiple parameters in a record class.
+* Command handlers only ever accept a single parameter and return effects. If a command handler needs to accept more than one parameter, those parameters should be wrapped inside a single command message.
+* The command handler parameter and reply should be defined as inner record of an entity, but sometimes it makes sense to place them in the `domain` package together with other data transfer and state records. Parameter and reply can also be a plain String, Java primitive or class.
+* Commands should be defined using an imperative naming convention, described as instructions of some task to be performed. For example, a command requesting a shopping cart be checked out would be `ShoppingCartEntity.Checkout`.
+* Command handlers are to be implemented in the entity and not in the state or domain object. Command handlers should invoke the domain object’s validators and methods that encapsulate the core business logic. The entity’s command handler is responsible for choosing the appropriate effect(s) to return. The domain object should not return effects or a list of events. This keeps the domain model clean of infrastructural concerns and makes it easy to unit test in isolation.
```

Comment:
This is probably the most difficult one. There are clearly two different styles/opinions about this. The alternative style that is not following the recommendation expressed here can be seen in choreography-saga

Reply:
Did the AI refactored it for you?

Reply:
I did that myself 😄

Reply:
You are not vibing enough!